src/main/java/de/hysky/skyblocker/skyblock/SmoothAOTE.java

```diff
@@ -0,0 +1,465 @@
+            if (offset != 0 && (isBlockFloor(checkPos.down()) || (isBlockFloor(checkPos.down().subtract(xDiagonalOffset)) && isBlockFloor(checkPos.down().subtract(zDiagonalOffset)))) && (pos.getY() - Math.floor(pos.getY())) < 0.31) {
+                closeFloorY = checkPos.getY() - 1;
+            }
+
+            //if the checking Y is same as closeY finish
+            if (closeFloorY == checkPos.getY()) {
+                return direction.multiply(offset - 1);
+            }
+        }
+
+        //return full distance if no collision found
+        return direction.multiply(distance);
+    }
+
+    /**
+     * Checks to see if a block is in the allowed list to teleport though
+     * Air, Buttons, carpets, crops, pots, mushrooms, nether wart, redstone, ladder, water, fire, lava, 3 or less snow layers
+     *
+     * @param blockPos block location
+     * @return if a block location can be teleported though
+     */
+    private static Boolean canTeleportThrough(BlockPos blockPos) {
+        if (CLIENT.world == null) {
+            return false;
+        }       
+
+        BlockState blockState = CLIENT.world.getBlockState(blockPos);
+        if (blockState.isAir()) {
+            return true;
+        }
+        Block block = blockState.getBlock();
+        return block instanceof ButtonBlock || block instanceof CarpetBlock || block instanceof CropBlock || block instanceof FlowerPotBlock || block.equals(Blocks.BROWN_MUSHROOM) || block.equals(Blocks.RED_MUSHROOM) || block.equals(Blocks.NETHER_WART) || block.equals(Blocks.REDSTONE_WIRE) || block.equals(Blocks.LADDER) || block.equals(Blocks.FIRE) || (block.equals(Blocks.SNOW) && blockState.get(Properties.LAYERS) <= 3) || block.equals(Blocks.WATER) || block.equals(Blocks.LAVA);
+    }
+
+    /**
+     * Checks to see if a block goes to the top if so class it as a floor
+     *
+     * @param blockPos block location
+     * @return if it's a floor block
+     */
+    private static Boolean isBlockFloor(BlockPos blockPos) {
+        if (CLIENT.world == null) {
+            return false;
+        }
+
+        BlockState blockState = CLIENT.world.getBlockState(blockPos);
+        VoxelShape shape = blockState.getCollisionShape(CLIENT.world, blockPos);
+        if (shape.isEmpty()) {
+            return false;
+        }
+        return shape.getBoundingBox().maxY == 1;
+    }
+
+    /**
+     * works out where they player should be based on how far though the predicted teleport time.
+     *
+     * @return the camera position for the interpolated pos
+     */
+
+    public static Vec3d getInterpolatedPos() {
+        if (CLIENT.player == null || teleportVector == null || startPos == null || teleportDisabled) {
+            return null;
+        }
+        long gap = System.currentTimeMillis() - startTime;
+        //make sure the player is actually getting teleported if not disable teleporting until they are teleported again
+        if (System.currentTimeMillis() - lastTeleportTime > 1000) {
```

Comment:
Actually, it might make sense to use ping here? Maybe two times so it's not too strict. Like this:
```suggestion
        if (System.currentTimeMillis() - lastTeleportTime > Math.min(2 * lastPing, MAX_TELEPORT_TIME)) {
```